### PR TITLE
Refactored top level dataflow main

### PR DESCRIPTION
**Pull request by iislucas**
Main pipeline code refactor: 
* Clarifying and simplified command line arguments (removed some duplicate argument specifications; and separated state paths from revisions path; clarified lots of variable names)
* Created a Locations class hold and manage the location paths. 
* Generalized paths so it will work on local filesystem and can be run locally without network (so that we can have local tests!). 
* Moved the reconstruction class into its own file.
* Added counters for more things and printing them out.
* Adjusted logging level to make the output readable.
* Made a simplified local_setup that doesn't have any special command and build magic (which I don't think is actually used, so I'll remove that later)
* Added example command line usage to top of the pipeline file.
* Added max_num_workers flag to example usage.

Other associated things: 
* Added testdata
* Fixed .gitignore to ignore temporary files. 
* Updated readme with setup instructions. 

Tested by running the command line at the top of the main pipeline file and manually inspecting the output files.

### CODE REVIEW

File: wikiconv/conversation_reconstruction/dataflow_main.py [hunk 1]
```diff
@@ -15,139 +15,230 @@
...
+  return (json.loads(s)["page_id"], json.loads(s))
+
+
+def index_by_rev_id(s):
+  """Given a json string of a dict with a rev_id key, pair the rev_id with
+  the string"""
+  return (json.loads(s)["rev_id"], s)
 
-  if known_args.testmode:
-    pipeline_args.append('--runner=DirectRunner')
```
Comment (vegetable68) [inline wikiconv/conversation_reconstruction/dataflow_main.py]: Might be good to keep an option to run as LocalRunner, as a different test choice.

File: wikiconv/conversation_reconstruction/dataflow_main.py [hunk 2]
```diff
@@ -15,139 +15,230 @@
...
+  --setup_file ./setup.py \
+  --input_state 'gs://YOUR_BUCKET/PATH_TO_INIT_STATE' \
+  --input_revisions 'gs://YOUR_BUCKET/PATH_TO_REVS*' \
+  --output_state 'gs://YOUR_BUCKET/PATH_TO_OUTPUT_STATE_TO' \
+  --output_conversations 'gs://YOUR_BUCKET/PATH_TO_OUTPUT_CONVS_TO' \
+  --runner DirectRunner
+  --project $YOUR_GCLOUD_PROJECT \
+  --job_name $NAME_OF_YOUR_JOB
+  --num_workers $NUMBER_OF_WORKERS_SUCH_AS_80
+```
```
Comment (vegetable68) [inline wikiconv/conversation_reconstruction/dataflow_main.py]: We need to change the parameters in helper shell accordingly.

Comment (vegetable68) [reply wikiconv/conversation_reconstruction/dataflow_main.py]: Also I think num_workers is "suggested number of workers", if we want to constrain it to max_num_workers, we need to change the code. I personally found max_num_workers was more useful.

File: wikiconv/conversation_reconstruction/dataflow_main.py (hunk 1 above)
Comment (iislucas) [reply wikiconv/conversation_reconstruction/dataflow_main.py]: The runner argument is now provided when we run the script (which assumes more dataflow knowledge (but we can document it more in the README), we we don't need this in the code as the person running the command can choose the runner themselves.

File: wikiconv/conversation_reconstruction/dataflow_main.py (hunk 2 above)
Comment (iislucas) [reply wikiconv/conversation_reconstruction/dataflow_main.py]: Thanks, I'll put in --max_num_workers into the suggested run command too.

**Comment by iislucas** [comment]
done PTAL and approve if you're happy. :)

**Pull request merged**